language/vm/src/abi.rs

```diff
@@ -0,0 +1,399 @@
+    Signer,
+    Struct {
+        address: AccountAddress,
+        module: Identifier,
+        name: Identifier,
+        type_arguments: Vec<TypeABI>,
+    },
+    Vector(Box<TypeABI>),
+    TypeParameter(TypeParameterIndex),
+    Reference(Box<TypeABI>),
+    MutableReference(Box<TypeABI>),
+}
+
+/// ABI for a `FieldDefinition`. The `name` is included even though it is
+/// metadata that it is ignored by the VM. The reason: names are important to clients. We would
+/// want a change from `Account { bal: u64, seq: u64 }` to `Account { seq: u64, bal: u64 }` to be
+/// marked as incompatible. Not safe to compare without an enclosing `StructABI`.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct FieldABI {
+    pub name: Identifier,
+    pub type_: TypeABI,
+}
+
+/// ABI for a kind constraint on a generic type parameter
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum KindABI {
+    All,
+    Resource,
+    Copyable,
+}
+
+/// ABI for a struct declaration. Not safe to compare without an associated `ModuleId` or `ModuleABI`.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct StructABI {
+    pub name: Identifier,
+    pub kind: KindABI,
+    pub type_parameters: Vec<KindABI>,
+    pub fields: Vec<FieldABI>,
+}
+
+/// ABI for a function declaration. Not safe to compare without an associated `ModuleId` or `ModuleABI`.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct FunctionABI {
+    pub name: Identifier,
+    pub type_parameters: Vec<KindABI>,
+    pub formals: Vec<TypeABI>,
+    pub ret: Vec<TypeABI>,
+}
+
+/// ABI for a `CompiledModule`: its address, name, struct declarations, and public function
+/// declarations.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct ModuleABI {
+    pub address: AccountAddress,
+    pub name: Identifier,
+    pub structs: Vec<StructABI>,
+    pub public_functions: Vec<FunctionABI>,
+}
+
+/// The result of an ABI compatibility check. Here is what the different combinations mean:
+/// `{ struct: true, struct_layout: true }`: fully backward compatible
+/// `{ struct_and_function_linking: true, struct_layout: false }`: Dependent modules that reference functions or types in this module may not link. However, fixing, recompiling, and redeploying all dependent modules will work--no data migration needed.
+/// `{ type_and_function_linking: true, struct_layout: false }`: Attempting to read structs published by this module will now fail at runtime. However, dependent modules will continue to link. Requires data migration, but no changes to dependent modules.
+/// `{ type_and_function_linking: false, struct_layout: false }`: Everything is broken. Need both a data migration and changes to dependent modules.
+pub struct Compatibility {
+    /// If false, dependent modules that reference functions or structs in this module may not link
+    pub struct_and_function_linking: bool,
+    /// If false, attempting to read structs previously published by this module will fail at runtime
+    pub struct_layout: bool,
+}
```

Comment:
I kinda get the desire of trying to be a little bit more specific on the exact kinds of incompatibility found, but I feel like this distinction is adding a lot more complexity to this story and I'm not sure if any client will  need to consume this information rather than a single bool. I also wonder if an `enum CompatibilityError` model will work better if we want to be specific.

Reply:
If you look at `stdlib/main.rs`, it is consuming the separate bools. Basically, we care about each case individually because it tells us what we need to do from a deployment perspective (redeploy code, or redeploy genesis)?

Agreed that if we were using this only for single-module upgrade, a single bool is sufficient/preferable. The `can_update: bool` API is intended for this use-case.